gateway/mw_streaming.go

```diff
@@ -0,0 +1,298 @@
+	sm.mw.Logger().Debugf("Creating stream: %s", streamFullID)
+
+	stream := streaming.NewStream(sm.mw.allowedUnsafe)
+	err := stream.Start(config, &handleFuncAdapter{mw: sm.mw, streamID: streamFullID, muxer: sm.muxer, sm: sm})
+	if err != nil {
+		sm.mw.Logger().Errorf("Failed to start stream %s: %v", streamFullID, err)
+		return err
+	}
+
+	sm.streams.Store(streamFullID, stream)
+	sm.mw.Logger().Infof("Successfully created stream: %s", streamFullID)
+
+	return nil
+}
+
+// ProcessRequest will handle the streaming functionality
+func (s *StreamingMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	strippedPath := s.Spec.StripListenPath(r.URL.Path)
+
+	s.Logger().Debugf("Processing request: %s, %s", r.URL.Path, strippedPath)
+
+	newRequest := r.Clone(r.Context())
+	newRequest.URL.Path = strippedPath
+
+	var match mux.RouteMatch
+	// First do the check if such route is defined by streaming API
+	//
+	// TODO: If we have a multiple streams, right now it will duplicate all this streams for each consumer
+	// If we have background jobs, or other non related streams, it will cause overhead and poential conflicts
+	// We need to meke .mux property of individual Steam object, and intiailize only the matched stream instead of all
+	if s.defaultStreamManager.muxer.Match(newRequest, &match) {
+		pathRegexp, _ := match.Route.GetPathRegexp()
+		s.Logger().Debugf("Matched stream: %v", pathRegexp)
+		handler, _ := match.Handler.(http.HandlerFunc)
+		if handler != nil {
+			// Now that we know that such streaming endpoint here,
+			// we can actually initialize individual stream manager
+			streamManager := s.createStreamManager(r)
+			streamManager.muxer.Match(newRequest, &match)
+
+			// direct Bento handler
+			handler, _ := match.Handler.(http.HandlerFunc)
+
+			handler.ServeHTTP(w, r)
+
+			// TODO: Implement shadowing
+			//
+			// if stream.Shaddow {
+			// 	go handler.ServeHTTP(w, r)
+			// 	return nil, http.StatusOK
+			// } else {
+			// 	handler.ServeHTTP(w, r)
+			// 	return nil, mwStatusRespond
+			// }
+
+			return nil, mwStatusRespond
+		}
+	}
+
+	// If no stream matches, continue with the next middleware
+	return nil, http.StatusOK
+}
+
+func (s *StreamingMiddleware) Unload() {
+	s.Logger().Debugf("Unloading streaming middleware %s", s.Spec.Name)
+
+	totalStreams := 0
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, _ interface{}) bool {
+			totalStreams++
+			return true
+		})
+		return true
+	})
+	globalStreamCounter.Add(-int64(totalStreams))
+
+	s.cancel()
+
+	s.Logger().Debug("Closing active streams")
+	s.streamManagers.Range(func(_, value interface{}) bool {
+		manager := value.(*StreamManager)
+		manager.streams.Range(func(_, streamValue interface{}) bool {
+			if stream, ok := streamValue.(*streaming.Stream); ok {
+				stream.Reset()
+			}
+			return true
+		})
+		return true
+	})
+
+	s.streamManagers = sync.Map{}
```

Comment:
Do we need to protect the fields of `StreamingMiddleware` struct with a lock? Members of the struct have been modified by different methods. How can be sure about thread-safety? It's hard and error prone to follow the execution flow with eye. We can use the race detector but it might be useful with an integration test that crafted for this purpose.

Reply:
Yeah agree, i will be changing it in the next Pr that comes with a refactor.